### ISSUE TITLE
Rename logp to logjoint and cache loglikelihood by default

### ISSUE THREAD

**Issue by mohamed82008** [TuringLang/Turing.jl#980]
I would like to propose the following change in `VarInfo`. I would like the `logp` field to be called `logjoint` and to have another field `loglikelihood` that caches the log likelihood by default and is then returned by the `Chain`. What do you think?

**Comment by willtebbutt** [comment]
Sounds reasonable, but what do you mean by log likelihood, as opposed to log joint?

**Comment by mohamed82008** [comment]
`log joint = log prior + log likelihood`

**Comment by mohamed82008** [comment]
After #965 we would be able to compute the likelihood only using the `LikelihoodContext`, this is useful for MLE for example. But for MCMC, we already compute the likelihood while sampling, we just don't store it. So storing it by default will allow the user to efficiently query the likelihood of the data given each parameter sample without having to recompute it.

**Comment by willtebbutt** [comment]
> log joint = log prior + log likelihood

Is the distinction between these things clear in an arbitrary programme?

**Comment by mohamed82008** [comment]
Hmm, I would define the log likelihood of the data given a set of parameters in an arbitrary Turing probabilistic programme as the `logp` that is produced by ignoring all the `assume` statements. That is every "random variable" or parameter is fixed to some value and doesn't contribute to the `logp`, only the `observe` functions contribute to the `logp`. From an implementation point of view, this quite clear. Do you think this is theoretically inaccurate?

**Comment by willtebbutt** [comment]
Okay, I see. So as I see it, there are two ways we could enable people to perform MLE.

1. The kind of thing that you're suggesting.
2. Via MAP with (usually) improper priors i.e. the user specifies that MAP should be performed.

The main upside of 1 is  a user can just specify that they want MLE to happen, and something MLE-like will happen without requiring them to modify your model. The benefit of 2 is that you avoid the main technical pitfall of 1 in that you avoid accidentally creating a hierarchical model in which you have some RVs that don't have any effect on the likelihood. e.g.
```julia
x ~ A
y ~ B(x)
z ~ C(y) # only make observations of this guy
```
In this model the likelihood is only a function of `y`, not `x`, so the values you get for `x` will be meaningless. It also seems likely that it will make whatever optimiser you are using behave weirdly, but I'm not really sure about that.

I don't think either approach is strictly superior to the other, but I would like to think about this before committing to a design decision.

**Comment by mohamed82008** [comment]
Beside MLE, computing the likelihood was also required for MLJ integration and to calculate things like LOO and WAIC (cc: @trappmartin and @sethaxen). Computing the log likelihood for a new observation or a new set of parameters is one thing and is already possible with #965, but this is not what I am proposing here. What I am proposing here is to store the already computed log likelihoods for the MCMC samples since we already have them. I think @trappmartin or @sethaxen might be able to comment better on whether this is useful or not.

**Comment by trappmartin** [comment]
Sounds like a very good idea to me.

Indeed, storing the likelihood values is necessary for most kinds of model selection, whether it is via WAIC or the model evidence. I'm not sure about the MLE use case.

But I feel that being able to store the log likelihood values for each observation & each sample would make sense and is in line with other PPLs (e.g. Stan has this feature and I believe PyMC too). But I would suggest that we don't store this information by default as it is only necessary in some cases.

**Comment by willtebbutt** [comment]
Fair enough, I'm happy now :)

**Comment by mohamed82008** [comment]
Hmm storing the likelihood of each observation and each sample will require another refactor of the `@preprocess` macro in #965 and a few other things. I will do it in another PR since that PR is already quite big.
